pandas/core/internals.py

```diff
@@ -2451,6 +2464,168 @@ def shift(self, periods, axis=0, mgr=None):
+        """ we internally represent the data as a PeriodIndex, but for external
+        compat with ndarray, export as a ndarray of Period """
+        return self.values.asobject.values
+
+    def to_dense(self):
+        # ToDo: FIXME
+        # called from Series.get_values()
+        return self.values.asobject
+
+    # def _try_fill(self, value):
+    #     """ if we are a NaT, return the actual fill value """
+    #     if value is tslib.NaT or np.array(isnull(value)).all():
+    #         value = tslib.iNaT
+    #     elif isinstance(value, Period):
+    #         # Period Nat can be handled here
+    #         value = value.ordinal
+    #     elif is_integer(value):
+    #         # regarded as ordinal
+    #         pass
+    #
+    #    return value
+
+    def _try_coerce_args(self, values, other):
+        """
+        Coerce values and other to dtype 'i8'. NaN and NaT convert to
+        the smallest i8, and will correctly round-trip to NaT if converted
+        back in _try_coerce_result. values is always ndarray-like, other
+        may not be
+
+        Parameters
+        ----------
+        values : ndarray-like
+        other : ndarray-like or scalar
+
+        Returns
+        -------
+        base-type values, values mask, base-type other, other mask
+        """
+        values_mask = isnull(values)
+        values = values.view('i8')
+        other_mask = False
+
+        if isinstance(other, bool):
+            raise TypeError
+        elif is_null_datelike_scalar(other):
+            other = tslib.iNaT
+            other_mask = True
+        elif isinstance(other, (Period)):
+            # ToDo: compare freq
+            other_mask = isnull(other)
+            other = other.ordinal
+        elif hasattr(other, 'dtype') and is_integer_dtype(other):
+            other = other.view('i8')
+        else:
+            try:
+                other = np.asarray(other)
+                other_mask = isnull(other)
+
+                other = other.astype('i8', copy=False).view('i8')
+            except ValueError:
+
+                # coercion issues
+                # let higher levels handle
+                raise TypeError
+
+        return values, values_mask, other, other_mask
+
+    def _try_coerce_result(self, result):
+        """ reverse of try_coerce_args / try_operate """
+        if isinstance(result, np.ndarray):
+            if is_integer_dtype(result):
+                result = self.values._shallow_copy(result)
+            else:
+                mask = isnull(result)
+                result = self.values._shallow_copy(result)
+                result[mask] = tslib.iNaT
+        elif isinstance(result, np.integer):
+            result = self._box_func(result)
+        return result
+
+    def should_store(self, value):
+        return isinstance(value, Period) and self.values.freq == value.freq
+
+    def to_native_types(self, slicer=None, na_rep=None,
+                        quoting=None, **kwargs):
```

Comment:
can also move (if use box_func) (though this IS datetimelike specific, but ok to stick in IndexHolder)